modules/standard/GMP.chpl

```diff
@@ -462,110 +451,142 @@ module GMP {
         halt("Error initializing big integer: bad format");
       }
+      owned = true;
     }
-    proc BigInt(str:string, base:int=0, out error:syserr) {
+    proc BigInt(str: string, base: int=0, out error:syserr) {
       var e:c_int;
       error = ENOERR;
       e = mpz_init_set_str(this.mpz, str.localize().c_str(), base.safeCast(c_int));
       if e {
         mpz_clear(this.mpz);
         error = EFORMAT;
       }
+      owned = true;
     }
-    proc BigInt(num:BigInt) {
-      if num.locale == here {
+    proc BigInt(ref num: BigInt, owned: bool = true) {
+      if num.locale_id == here.id {
         mpz_init_set(this.mpz, num.mpz);
       } else {
         mpz_init(this.mpz);
-        var mpz_struct = num.mpzStruct(); 
+        var mpz_struct = num.mpzStruct();
         chpl_gmp_get_mpz(this.mpz, num.locale.id, mpz_struct);
       }
+      this.owned = owned;
     }
-    proc BigInt() { mpz_init(this.mpz); }
 
     // destructor
-    proc ~BigInt() { on this do mpz_clear(this.mpz); }
+    proc ~BigInt(){
+        if this.owned then {
+          on Locales[this.locale_id] do mpz_clear(this.mpz);
+          this.owned = false;
+        } //else do nothing
+    }
+
+
+    // Note: by the time reinit is called, locality has been taken care of
+    proc ref reinitBigInt(num : mpz, needToCopy : bool = true) {
+      // if the record exists, and is owned, a constructor of some
+      // sort was called, and mpz is initialized. 
+
+      // if we own our old num re-use it
+      // if we don't, init a new one
+      if !(this.owned) { //TODO: confirm that this case gets tested
+        on Locales[this.locale_id] do mpz_init(this.mpz);
+        this.owned = true;
+      }
+
+      // if we don't need to deep copy, free ours, make a shallow copy, get out
+      if !needToCopy {
+        on Locales[this.locale_id] do mpz_clear(this.mpz);
+        this.owned = false;
+        this.mpz = num;
+        return;
+      }
+      else {
+        mpz_set(this.mpz, num);
+        this.owned = true;
+      }
+    }
 
     // utility functions used below.
-    proc numLimbs:uint(64) {
+    proc numLimbs: uint(64) {
       var mpz_struct = this.mpz[1];
       return chpl_gmp_mpz_nlimbs(mpz_struct);
     }
 
-    proc mpzStruct():__mpz_struct {
-      var ret:__mpz_struct;
-      on this {
+    proc mpzStruct(): __mpz_struct {
+      var ret: __mpz_struct;
+      on Locales[this.locale_id] {
         ret = this.mpz[1];
       }
       return ret;
     }
 
-
-    // returns true if we made a temp copy.
-    proc maybeCopy():(bool,BigInt) {
-      if this.locale == here {
-        return (false,this);
-      } else {
-        var mpz_struct = this.mpz[1];
-        var tmp = new BigInt(true, (mp_bits_per_limb:uint(64))*chpl_gmp_mpz_nlimbs(mpz_struct));
-        chpl_gmp_get_mpz(tmp.mpz, this.locale.id, mpz_struct); 
-        return (true, tmp);
-      }
-    }
-
-    // Assignment functions
-    proc set(a:BigInt)
+   // Assignment functions
```

Comment:
Probably a topic for face-to-face

I am not sure what is happening to the existing state of the current BigInt for owned == true vs owned == false